### PR TITLE
MAKES SPIFFS Great (Again)

### PR DESCRIPTION
**Pull request by cjbrigato**
SPIFFS : FIXES FS corruption against high number of current losses, high number of files manipulation. FIXES LOG_BLOCK erasing handlers. FIXES Garbage Collector. Overall, SPIFFS should now be WAY MORE resilient to any form of misbehavior. In case of corrupted filesystem (should still be readable !), spamming 'mem spiffs check' from the client will ensure at least 2 new free 4k blocks each tile (if number of LOG_PAGE available is enough) and in the same pass will move/reorganize the LOG_PAGES of contextual blocks. Yiha

### CODE REVIEW

**Comment by cjbrigato** [comment]
@doegox woops, this is result of multiple cherrypick of my branches here and there, it MAY need adding your new includes restructuration again inside. 
(if possible, avoid stddef anywhere near SPIFFS internals)

**Comment by doegox** [comment]
@cjbrigato ok I'll see for the merge

File: armsrc/spiffs.c [hunk 1]
```diff
@@ -533,21 +539,16 @@ void rdv40_spiffs_safe_print_fsinfo() {
...
-    Dbprintf("  Max Path Length............" _YELLOW_("%d")"chars", fsinfo.maxPathLength);
-//    DbpString(_BLUE_("Details"));
-    DbpString("");
-    Dbprintf("  Filesystem\tSize\tUsed\tAvailable\tUse%\tMounted on");
-    Dbprintf("  spiffs    \t%d B\t%d B\t%d B\t\t"_YELLOW_("%d%")"\t/"
-             , fsinfo.totalBytes
-             , fsinfo.usedBytes
-             , fsinfo.freeBytes
-             , fsinfo.usedPercent
-            );
```
Comment (iceman1001) [inline armsrc/spiffs.c]: the "\t" tabs works very bad on WSL.   I strongly suggest consistent use of 4 spaces instead of tab.

**Comment by doegox** [comment]
I'm merging locally and fixing issues, please don't add commits on this PR at the same time :)

File: armsrc/spiffs.c (hunk 1 above)
Comment (cjbrigato) [reply armsrc/spiffs.c]: Are you seriously using WSL as an argument in about how should TTy-formating be done ?
:D 
More seriously, Feel free to revert back some change that my cherrypicks may have lost and to show how consistant space allways had been through space and time.
woooossh

**Comment by cjbrigato** [comment]
This PR won't have any more commit, I knew I was far far away from any correct git flow, so I thought you will be pleased by a 1-commit PR.

**Comment by doegox** [comment]
> if possible, avoid stddef anywhere near SPIFFS internals

spiffs_config.h  is including string.h which needs size_t, how to solve that without stddef?
(and that's just an example, size_t is everywhere in arm code)

File: armsrc/spiffs.c (hunk 1 above)
Comment (iceman1001) [reply armsrc/spiffs.c]: .... yes....   because I am naughty.

There is some string format woodoo to get fixed cols instead of using tabs.

**Comment by cjbrigato** [comment]
@doegox this is very strange but when stddef come from common.h and nowhere else in between I can't make the bug again. May it have been some missing endifness or some git-ignored Colinism ? Anyway, stddefGate is nonevent finally .

**Comment by doegox** [comment]
so we can use stddef?

**Comment by cjbrigato** [comment]
_stddef_initely

**Comment by doegox** [comment]
ok playing with kid now, I'll finish it later, and include std stuff directly rather than through common.h, for clarity

**Comment by cjbrigato** [comment]
Yeah just to be clear the PR is absolutely not about the restruct thing. Considering the rehab of stddef, it should just work as-is with new structure.
In fact it's good thing it triggered so it revealed some weaknesses here and there, enough to be nearly-unrecoverable for filesystems.

This PR should make everything way more *proof. And if anyhow someone is in need of unusual reliability (in which case someone should not rely on this flash and filesystem setup, anyway) Enabling PAGE_CHECK should do the trick.
Disabling Write cache is more tricky. Write-cached loss is juste some write loss.
A Direct-Write loss may be from benign corruption to fatal to data.

**Comment by doegox** [comment]
merged manually and fixed includes

File: armsrc/spiffs.c (hunk 1 above)
Comment (doegox) [reply armsrc/spiffs.c]: FYI there is a ton of files with such printf \t
./client/crypto/asn1dump.c
./client/proxguiqt.cpp
./client/mifare/ndef.c
./client/mifare/mifarehost.c
./client/reveng/cli.c
./client/reveng/poly.c
./client/reveng/bmpbit.c
./client/jansson/dump.c
./client/jansson/load.c
./client/jansson/pack_unpack.c
./client/loclass/hash1_brute.c
./client/loclass/cipherutils.c
./client/emv/emv_tags.c
./client/emv/test/cryptotest.c
./client/emv/cmdemv.c
./client/emv/dump.c
./client/cmdlf.c
./client/cliparser/cliparser.c
./client/cliparser/argtable3.c
./client/cmdhfmfp.c
./client/cmdhficlass.c
./client/tinycbor/cborpretty.c
./client/scripts/Legic_clone.lua
./client/scripts/ntag_3d.lua
./client/scripts/legic.lua
./client/cmdhfmfu.c
./client/lualibs/default_toys_di.lua
./client/lualibs/dkjson.lua
./client/lualibs/default_toys.lua
./client/hardnested/hardnested_bruteforce.c
./client/cmdhffelica.c
./client/pm3_mfdread.py
./client/flasher.c
./client/cmdlfindala.c
./client/cmdhf15.c
./client/cmdsmartcard.c
./client/cmdhffido.c
./client/cmdhfmf.c
./client/liblua/lauxlib.c
./client/liblua/lbaselib.c
./client/liblua/llex.c
./client/liblua/luac.c
./client/liblua/loadlib.c
./client/cmdhf14b.c
./client/cmdhf14a.c
./client/util.c
./armsrc/spiffs_hydrogen.c
./armsrc/spiffs.c
./armsrc/frozen.c
./armsrc/Standalone/hf_colin.c
./armsrc/Standalone/hf_mattyrun.c
./fpga/tests/tb_lf_edge_detect.v
./fpga/tests/tb_min_max_tracker.v
./tools/deprecated-hid-flasher/flasher/flasher.c
./tools/deprecated-hid-flasher/flasher/proxusb.c
./tools/analyzesize.py
./tools/xorcheck.py
./tools/findbits.py
./common/mbedtls/bn_mul.h

**Pull request closed without merging**